grpc_transcode.go

```diff
@@ -0,0 +1,602 @@
+		if !w.wroteHeader {
+			w.writeHeader()
+		}
+		return w.ResponseWriter.Write(b)
+	}
+
+	prefixSize := 5 - w.wrotePrefix
+	if prefixSize <= 0 {
+		n, err = w.body.Write(b)
+	} else {
+		size := len(b)
+		if size < prefixSize {
+			prefixSize = size
+		}
+		w.wrotePrefix += prefixSize
+		b = b[prefixSize:]
+		n, err = w.body.Write(b)
+	}
+	total := w.body.Len() + w.wrotePrefix
+	limit := w.config.SendMaxBytes
+	if limit > 0 && total > limit {
+		return n, NewError(CodeResourceExhausted, fmt.Errorf("message size %d exceeds sendMaxBytes %d", total, limit))
+	}
+	return n, err
+}
+
+func getGRPCTrailer(header http.Header) http.Header {
+	isTrailer := map[string]bool{
+		// Force GRPC status values, try copy them directly from header.
+		grpcHeaderStatus:  true,
+		grpcHeaderDetails: true,
+		grpcHeaderMessage: true,
+	}
+	for _, key := range strings.Split(header.Get(headerTrailer), ",") {
+		key = http.CanonicalHeaderKey(key)
+		isTrailer[key] = true
+	}
+
+	trailer := make(http.Header)
+	for key, vals := range header {
+		key = http.CanonicalHeaderKey(key)
+		if strings.HasPrefix(key, http.TrailerPrefix) || isTrailer[key] {
+			key = strings.TrimPrefix(key, http.TrailerPrefix)
+			trailer[key] = vals
+		}
+	}
+	return trailer
+}
+
+func (w *connectResponseWriter) finalize() error {
+	if !w.wroteHeader {
+		w.writeHeader()
+	}
+
+	header := w.Header()
+	trailer := getGRPCTrailer(header)
+
+	protobuf := w.config.Codecs[codecNameProto]
+	trailerErr := grpcErrorFromTrailer(w.config.BufferPool, protobuf, trailer)
+	if trailerErr != nil {
+		if w.isStreaming() {
+			trailerErr.meta = trailer
+		}
+		return trailerErr
+	}
+
+	// Remove all gRPC trailer keys.
+	for key := range trailer {
+		if isGRPCHeader[key] {
+			delete(trailer, key)
+		}
+	}
+
+	if w.isStreaming() {
+		// Encode as connect end message.
+		end := &connectEndStreamMessage{
+			Trailer: trailer,
+		}
+		data, err := json.Marshal(end)
+		if err != nil {
+			panic(err)
+		}
+
+		head := []byte{connectFlagEnvelopeEndStream, 0, 0, 0, 0}
+		binary.BigEndian.PutUint32(head[1:5], uint32(len(data)))
+		if _, err := w.ResponseWriter.Write(head); err != nil {
+			return err
+		}
+		if _, err := w.ResponseWriter.Write(data); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// Encode trailers as header
+	responseHeader := w.ResponseWriter.Header()
+	for key, vals := range trailer {
+		if isGRPCHeader[key] || isConnectHeader[key] {
+			continue
+		}
+		key = connectUnaryTrailerPrefix + key
+		responseHeader[key] = vals
+	}
+	// Write buffered unary body to stream.
+	if _, err := w.body.WriteTo(w.ResponseWriter); err != nil {
+		return err
+	}
+	flushResponseWriter(w.ResponseWriter)
+	return nil
+}
+
+func (w *connectResponseWriter) Flush() {
+	if w.isStreaming() {
+		flushResponseWriter(w.ResponseWriter)
+	}
+}
+
+// GRPCHandler translates connect and gRPC-web to a gRPC request for
+// use with gRPC handlers.
+func GRPCHandler(h http.Handler, options ...HandlerOption) http.Handler {
```

Comment:
The use of `HandlerOption` here feels a bit misleading since most options don't apply.

Perhaps instead we need a new `MiddlewareOption` interface that extends (embeds) `HandlerOption` and use it just for the options it cares about (which seems to be max read amount and whether the connect version header is required). One trick is that max read is used for both handler and client -- so we might need yet another, like `OptionAny` that implements both `Option` and `MiddlewareOption`.